### PR TITLE
[Xiangzhe Zhang]:Fix bug

### PR DESCRIPTION
**Pull request by JiaoTuan**
The way to look up the kallsyms table should be to use binary search to find the target address or the address closest to the target address

### CODE REVIEW

**Comment by JiaoTuan** [comment]
When I used bpf_get_stackid to capture the function call stack, I found that the function address I got did not correspond to kallsyms. I modified the function with reference to the ksym_search function.

File: helpers/kernel_symbols.go [hunk 1]
```diff
@@ -101,12 +101,38 @@ func (k *KernelSymbolTable) GetSymbolByName(owner string, name string) (*KernelS
...
+	var keys []uint64
+	for _, Symbol := range k.symbolMap {
+		keys = append(keys,Symbol.Address)
+	}
+	sort.Slice(keys, func(i, j int) bool { return keys[i] < keys[j] })
+	kallsyms_addr :=bin_search(keys,addr)
+	
 	for _, Symbol := range k.symbolMap {
-		if Symbol.Address == addr {
+		if Symbol.Address == kallsyms_addr {
```
[the remark this answers is not in the document]
Comment (javierhonduco) [reply helpers/kernel_symbols.go:133]: I might be misunderstanding something, but I think we need to perform a binary search as we could get any address in between functions, something like:

| Start address      | Function |
| ----------- | ----------- |
| 0      | first_func       |
| 100   | second_func        |
| 900   | third_func        |

For the instruction pointer=120, for example, the function would be `second_func` as it's base address is lower than 120 and the next function's base address is higher, but not equal

File: helpers/kernel_symbols.go [hunk 2]
```diff
@@ -101,12 +101,38 @@ func (k *KernelSymbolTable) GetSymbolByName(owner string, name string) (*KernelS
...
+}
+
 // GetSymbolByAddr returns a symbol by a given address
 func (k *KernelSymbolTable) GetSymbolByAddr(addr uint64) (*KernelSymbol, error) {
+	var keys []uint64
+	for _, Symbol := range k.symbolMap {
+		keys = append(keys,Symbol.Address)
+	}
+	sort.Slice(keys, func(i, j int) bool { return keys[i] < keys[j] })
+	kallsyms_addr :=bin_search(keys,addr)
```
[the remark this answers is not in the document]
Comment (rafaeldtinoco) [reply helpers/kernel_symbols.go:130]: @grantseltzer I believe we should run similar checks from what we currently have in tracee: https://github.com/aquasecurity/tracee/blob/main/.github/workflows/pr.yaml#L40 on each PR.

File: helpers/kernel_symbols.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (rafaeldtinoco) [reply helpers/kernel_symbols.go:133]: Totally fine, just so Im clear, I was the one who misread the original intent =D.

**Comment by grantseltzer** [comment]
Its been a while since this has been updated and it seems we haven't come to any conclusions about whether or not this is the right change. I believe we should keep the function such that only a direct match of address should resolve the symbol. As such i'm closing this but fully welcome another PR to further the discussion @JiaoTuan

**Pull request closed without merging**